cmd/tempo-vulture/main.go

```diff
@@ -45,95 +43,189 @@ func init() {
+	flag.DurationVar(&tempoRetentionDuration, "tempo-retention-duration", 336*time.Hour, "The block retention that Tempo is using")
 }
 
 func main() {
 	flag.Parse()
 
-	glog.Error("Application Starting")
+	glog.Error("Tempo Vulture Starting")
 
-	testDurations := []time.Duration{
-		24 * time.Hour,
-		12 * time.Hour,
-		6 * time.Hour,
-		3 * time.Hour,
-		time.Hour,
-		30 * time.Minute,
-	}
+	startTime := time.Now().Unix()
+	ticker := time.NewTicker(tempoBackoffDuration)
+	slot := int64(tempoBackoffDuration/time.Second) * 2
 
-	ticker := time.NewTicker(15 * time.Second)
+	// Write
 	go func() {
 		for {
 			<-ticker.C
 
-			for _, duration := range testDurations {
+			rand.Seed((time.Now().Unix() / slot) * slot)
+			c, err := newJaegerGRPCClient(tempoPushURL)
+			if err != nil {
+				glog.Error("error creating grpc client", err)
+				metricErrorTotal.Inc()
+				continue
+			}
 
-				// query loki for trace ids
-				lines, err := queryLoki(lokiBaseURL, lokiQuery, duration, lokiUser, lokiPass)
+			traceID := rand.Int63()
+			for i := int64(0); i < generateRandomInt(1, 100); i++ {
+				err = c.EmitBatch(context.Background(), makeThriftBatch(traceID))
 				if err != nil {
-					glog.Error("error querying Loki ", err)
+					glog.Error("error pushing batch to Tempo ", err)
 					metricErrorTotal.Inc()
 					continue
 				}
-				ids := extractTraceIDs(lines)
+			}
+		}
+	}()
 
-				// query tempo for trace ids
-				metrics, err := queryTempoAndAnalyze(tempoBaseURL, tempoBackoffDuration, ids)
-				if err != nil {
-					glog.Error("error querying Tempo ", err)
-					metricErrorTotal.Inc()
-					continue
-				}
+	// Read
+	go func() {
+		for {
+			<-ticker.C
+
+			currentTime := time.Now().Unix()
 
-				metricTracesInspected.WithLabelValues(strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.requested))
-				metricTracesErrors.WithLabelValues("notfound", strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.notfound))
-				metricTracesErrors.WithLabelValues("missingspans", strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.missingSpans))
+			// don't query traces before retention
+			if (currentTime - startTime) > int64(tempoRetentionDuration/time.Second) {
+				startTime = currentTime - int64(tempoRetentionDuration/time.Second)
 			}
+
+			// pick past slot and re-generate trace
+			rand.Seed((generateRandomInt(startTime, currentTime) / slot) * slot)
+			traceID := rand.Int63()
+			batch := makeThriftBatch(traceID)
+			hexID := fmt.Sprintf("%016x%016x", batch.Spans[0].TraceIdHigh, batch.Spans[0].TraceIdLow)
+
+			// query the trace
+			metrics, err := queryTempoAndAnalyze(tempoQueryURL, hexID)
+			if err != nil {
+				glog.Error("error querying Tempo ", err)
+				metricErrorTotal.Inc()
+				continue
+			}
+
+			metricTracesInspected.Add(float64(metrics.requested))
+			metricTracesErrors.WithLabelValues("notfound").Add(float64(metrics.notfound))
```

Comment:
let's add a "failed" metric for anything that's not 404 or 200.

Reply:
I'm re-using `util.QueryTrace`, so I added this new metric when that function returns an error and it's not `util.ErrTraceNotFound`.

I increment both `metricErrorTotal` and `metricTracesErrors{error="failed"}`, that's ok?